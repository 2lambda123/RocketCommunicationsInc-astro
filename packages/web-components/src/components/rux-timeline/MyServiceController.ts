class MyServiceController {
    public myData = []
    constructor() {}
    async load() {
        if (this.myData) {
            return this.myData
        } else {
            // Load data and then...
            return this.myData
        }
    }
    async getData() {
        const data = await this.load()
        return data
    }
    addData(data: any) {
        //@ts-ignore
        this.myData.push(data)
    }
}
export const MyService = new MyServiceController()
