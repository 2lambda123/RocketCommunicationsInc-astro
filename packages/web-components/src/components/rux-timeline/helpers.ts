import {
    addHours,
    differenceInHours,
    eachHourOfInterval,
    addDays,
    addMinutes,
    subMinutes,
    differenceInDays,
    differenceInMonths,
    eachDayOfInterval,
    parseISO,
} from 'date-fns'
import { formatInTimeZone, utcToZonedTime, zonedTimeToUtc } from 'date-fns-tz'

export async function validateTimezone(timezone: string) {
    return new Promise((resolve, reject) => {
        try {
            formatInTimeZone(new Date(), timezone, 'MM/dd')
            resolve(true)
        } catch (e) {
            reject(false)
        }
    })
}

// Fixes bug in date-fn's subDays() function when crossing a DST transition:
// https://github.com/date-fns/date-fns/issues/571

function agnosticAddDays(date: Date, amount: number) {
    const originalTZO = date.getTimezoneOffset()
    const endDate = addDays(date, amount)
    const endTZO = endDate.getTimezoneOffset()

    const dstDiff = originalTZO - endTZO

    return dstDiff >= 0
        ? addMinutes(endDate, dstDiff)
        : subMinutes(endDate, Math.abs(dstDiff))
}

export function dateRange(
    start: any,
    end: any,
    interval: any,
    intervalValue: any = 1,
    timezone: any = 'UTC'
) {
    const startDate = new Date(start)
    const endDate = new Date(end)

    // Prevents unnecessary error when setting range to the future
    if (startDate > endDate) {
        return []
    }

    if (interval === 'day') {
        const days = differenceInDays(endDate, startDate)

        const output = [...Array(days).keys()].map((i) => {
            const time = agnosticAddDays(startDate, i)

            const formattedTime = formatInTimeZone(time, timezone, 'MM/dd')

            return formattedTime
        })

        return output
    }

    if (interval === 'hour') {
        // intervalValue = 2
        let days = differenceInHours(endDate, startDate)
        console.log('days', days)
        console.log('intervaldays', days / intervalValue)
        days = days / intervalValue

        // const newout = [...Array(days).keys()].reduce((a: any, b: any) => {

        //     let offset = 0
        //     if (a.length === 0 ) {

        //     }
        //     if (a.length > 1) {
        //         const last = a[a.length-1]
        //         // if (last) {
        //         //     const hour = parseInt(last.slice(0,2))
        //         //     offset = hour
        //         //     console.log('off', last);
        //         // }
        //         offset = last
        //     }

        //     // this might break bc new date not wrapped in timezone
        //     const time = addHours(offset, 2)

        //     const formattedTime = formatInTimeZone(time, timezone, 'HH:mm')
        //     a.push(formattedTime)
        //     return a

        //     // console.log('a',a);
        //     // console.log('b',b);

        // }, ['00:00'])
        // console.log('new', newout);

        console.log('start', start)

        const eachDay = eachHourOfInterval(
            {
                start: new Date(start),
                end: new Date(end),
            },
            {
                step: 2,
            }
        )
        const test = eachDay.map((day) => {
            return formatInTimeZone(day, timezone, 'HH:mm')
        })
        return test

        // const output = [...Array(days).keys()].map((i) => {

        //     const time = addHours(startDate, i + intervalValue)

        //     const formattedTime = formatInTimeZone(time, timezone, 'HH:mm')
        //     return formattedTime
        // })
        // console.log('out', output);

        // return output
    }

    return []
}

export function dateRangeInMonths(
    start: any,
    end: any,
    interval: any,
    intervalValue: any = 1,
    timezone: any = 'UTC'
) {
    const startDate = new Date(start)
    const endDate = new Date(end)

    // Prevents unnecessary error when setting range to the future
    if (startDate > endDate) {
        return []
    }

    if (interval === 'day') {
        const days = differenceInDays(endDate, startDate)

        const output = [...Array(days).keys()].map((i) => {
            const time = agnosticAddDays(startDate, i)

            const formattedTime = formatInTimeZone(time, timezone, 'MM/dd')

            return formattedTime
        })

        return output
    }

    if (interval === 'hour') {
        let days = differenceInHours(endDate, startDate)
        let months = differenceInDays(endDate, startDate)

        // days = days / intervalValue

        // const output = [...Array(days).keys()].map((i) => {
        //     const time = addHours(startDate, i)

        //     const formattedTime = formatInTimeZone(time, timezone, 'HH:mm')
        //     return formattedTime
        // })

        const output = [...Array(months).keys()].map((i) => {
            const time = agnosticAddDays(startDate, i)

            const formattedTime = formatInTimeZone(time, timezone, 'MM/dd')

            return formattedTime
        })
        console.log('out', output)

        return output
    }

    return []
}
