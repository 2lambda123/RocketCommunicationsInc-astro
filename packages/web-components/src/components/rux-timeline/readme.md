# rux-timeline

<!-- Auto Generated Below -->

## Properties

| Property   | Attribute   | Description | Type     | Default                  |
| ---------- | ----------- | ----------- | -------- | ------------------------ |
| `end`      | `end`       |             | `string` | `'2021-02-10T00:00:00Z'` |
| `start`    | `start`     |             | `string` | `'2021-02-01T00:00:00Z'` |
| `totalCol` | `total-col` |             | `any`    | `null`                   |
| `zoom`     | `zoom`      |             | `number` | `120`                    |

## Methods

### `getTotalColumns() => Promise<number>`

#### Returns

Type: `Promise<number>`

---

_Built with [StencilJS](https://stenciljs.com/)_
