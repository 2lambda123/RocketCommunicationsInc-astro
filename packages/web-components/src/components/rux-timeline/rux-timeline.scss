:host {
    display: block;
    ::-webkit-scrollbar {
        width: 16px;
        height: 16px;
        margin-left: -20px;
        background-color: transparent;
    }

    ::-webkit-scrollbar-thumb {
        background-color: var(--color-active, rgb(43, 101, 155));
        border-radius: 8px;
        border: 3px solid transparent;
        background-clip: padding-box;
    }

    /* visually "centers" because the dark edge of the shadow gives the illusion this is offset */
    ::-webkit-scrollbar-thumb:vertical {
        border-left-width: 4px;
    }

    ::-webkit-scrollbar-thumb:horizontal {
        border-top-width: 4px;
    }

    ::-webkit-scrollbar-thumb:active,
    ::-webkit-scrollbar-thumb:hover {
        background-color: var(--color-primary, rgb(58, 129, 191));
    }

    ::-webkit-scrollbar-track,
    ::-webkit-scrollbar-corner {
        background-color: var(--color-surface, rgb(27, 45, 62));
    }

    ::-webkit-scrollbar-track:vertical {
        box-shadow: inset 3px 3px 3px 0px rgba(0, 0, 0, 0.5);
    }

    ::-webkit-scrollbar-track:horizontal {
        box-shadow: inset 1px 3px 3px 0px rgba(0, 0, 0, 0.5);
    }
}

.rux-timeline {
    // scrollbar-color: var(--color-active);
    // overflow: auto;
    // position: relative;
    // display: grid;
    // grid-template-rows: auto;
    // background: var(--color-surface);
    display: flex;
    flex-direction: column;
    overflow-x: auto;
}

.rux-timeline-overlay {
    scrollbar-color: var(--color-active);

    overflow: auto;
    position: relative;
    display: grid;
    // grid-gap: 1em;
    // grid-template-columns: [header] 200px repeat(12, 120px);
    // [time-0000] 120px
    // [time-0100] 120px
    // [time-0200] 120px
    // [time-0300] 120px
    // [time-0400] 120px
    // [time-0500] 120px
    // [time-0600] 120px
    // [time-0700] 120px
    // [time-0800] 120px
    // [time-0900] 120px
    // [time-1000] 120px
    // [time-1100] 120px
    // [time-1200] 120px
    // [time-1300] 120px
    // [time-1400] 120px
    // [time-1500] 120px
    // [time-1600] 120px
    // [time-1700] 120px
    // [time-1800] 120px
    // [time-1900] 120px
    // [time-2000] 120px
    // [time-2100] 120px
    // [time-2200] 120px
    // [time-2300] 120px;
    // [time-2359];

    grid-template-rows: auto;
    background: var(--color-surface);
}

// .rux-playhead {
//     width: 5px;
//     //   grid-row: 1 / 5;
//     //   grid-column: time-0000 / time-2359;

//     background: var(--color-primary);
//     z-index: 10;

//     height: 100%;
//     position: absolute;
//     top: 0;
//     // left: 200px;
//     &:after {
//         width: 0;
//         height: 0;
//         border-left: 12px solid transparent;
//         border-right: 12px solid transparent;
//         border-top: 12px solid var(--color-primary);
//         content: '';
//         position: absolute;
//         left: -9.5px;
//     }
// }

.hidden {
    display: none;
}

.cell {
    // border: 1px solid blue;
    border-right: 0;
    border-top: 0;
    // grid-column: span 60;
    z-index: 9999;
    grid-column: span 1;
}
.cell:nth-child(10n) {
    // background: blue;
}

.cell:nth-child(odd) {
    // background: var(--color-background);
    background: red;
}

.cell:nth-child(even) {
    background: blue;
    // opacity: 10%;
}

.marker {
    background: blue;
}

.rux-playhead {
    width: 2px;
    //   grid-row: 1 / 5;
    //   grid-column: time-0000 / time-2359;

    background: var(--color-primary);
    z-index: 10;

    height: 100%;
    position: absolute;
    top: 0;
    // left: 200px;
    &:after {
        width: 0;
        height: 0;
        border-left: 12px solid transparent;
        border-right: 12px solid transparent;
        border-top: 12px solid var(--color-primary);
        content: '';
        position: absolute;
        left: -11.5px;
    }
}
