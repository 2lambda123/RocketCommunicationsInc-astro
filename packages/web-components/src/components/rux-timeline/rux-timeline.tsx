import {
    Watch,
    Element,
    State,
    Method,
    Component,
    Host,
    h,
    Prop,
} from '@stencil/core'
import { format, parse } from 'date-fns'
import { utcToZonedTime } from 'date-fns-tz/esm'
import { addMinutes, differenceInHours } from 'date-fns/esm'
import differenceInMinutes from 'date-fns/esm/fp/differenceInMinutes/index.js'
import { RuxIconContactSupport } from '../rux-icon/icons/rux-icon-contact-support'
import { dateRange } from './helpers'

@Component({
    tag: 'rux-timeline',
    styleUrl: 'rux-timeline.scss',
    shadow: true,
})
export class RuxTimeline {
    private slotContainer?: HTMLElement
    public slots?: any = 'empty'
    @State() newTime: any = ''
    @State() margin = 200
    @State() time = '00:00'
    @Element() el!: HTMLRuxTimelineElement
    @Prop() start = '2021-02-01T00:00:00Z'
    @Prop() end = '2021-02-10T00:00:00Z'
    @Prop() totalCol: any = null
    @Prop() zoom = 120
    @Prop() interval: 'hour' | 'day' | 'month' = 'hour'

    @Watch('zoom')
    handleZoomChange(old: any, newValue: any) {
        const newMargin = this.calcPlayheadFromTime(this.newTime)
        this.margin = newMargin
    }

    @Watch('start')
    @Watch('end')
    handleStartChange() {
        this.calcDiff()
    }
    connectedCallback() {
        this._handleSlotChange = this._handleSlotChange.bind(this)
        this.handleMouse = this.handleMouse.bind(this)
    }
    componentWillLoad() {
        const childNodes = this.el.childNodes
        const children = Array.prototype.filter.call(
            childNodes,
            (node) => node.nodeType == Node.ELEMENT_NODE
        )

        children.forEach((el, index) => {
            //@ts-ignore
            el.track = ++index
        })
        // this._handleSlotChange()

        window.setInterval(() => {
            // this.margin = this.margin + 2
            //   console.log(this.margin);
        }, 1000)

        // const { width } = this.el.getBoundingClientRect()
    }

    @Method()
    async getTotalColumns() {
        const start = new Date(this.start)
        const end = new Date(this.end)
        return differenceInHours(start, end)
    }

    calcDiff() {
        const test = dateRange(this.start, this.end, this.interval)
        this.totalCol = test.length
    }

    get ratio() {
        if (this.interval === 'hour') {
            return this.zoom / 60 // for hours.
        }

        if (this.interval === 'day') {
            return this.zoom / 120 //tbd
        }
        return 2
    }

    calcTimeFromPlayhead(position: any) {
        this.margin = position

        const time = position - 200

        const min = time / this.ratio

        let intervalValue = 60
        if (this.interval === 'day') {
        }
        // const start = utcToZonedTime(this.start, 'utc')
        const newTime = addMinutes(new Date(this.start), min)
        const newTimeFormatted = format(newTime, 'MM/dd/Y HH:mm:ss')
        this.newTime = newTime

        const hours = Math.floor(min / intervalValue)
        const minutes = Math.floor(min % intervalValue)

        return `${hours}:${minutes}`
    }

    calcPlayheadFromTime(time: any, ratio?: any) {
        if (!ratio) {
            ratio = this.ratio
        }

        // const start = utcToZonedTime(this.start, 'utc')
        // const targetTime = utcToZonedTime(time, 'utc')
        const newTime = differenceInMinutes(
            new Date(this.start),
            new Date(time)
        )

        const result = newTime * ratio + 200

        return result
    }

    handleMouse(e: any) {
        const rect = this.el.getBoundingClientRect()
        const scrollOffset = this.slotContainer
            ? this.slotContainer?.scrollLeft
            : 0

        const position = e.clientX - rect.left + scrollOffset

        // if (e.clientY <= 234) { // ignore scrollbar

        if (position > 200) {
            this.time = this.calcTimeFromPlayhead(position)
            // this.calcPlayheadFromTime(this.time)
        } else {
            this.margin = 200
        }
        // }
    }
    get theSlots() {
        return this.slots
    }

    get formattedCurrentTime() {
        if (this.newTime) {
            return format(this.newTime, 'MM/dd/Y HH:mm:ss')
        } else {
            return null
        }
    }

    private _handleSlotChange() {
        const slot = this.slotContainer?.querySelector(
            'slot'
        ) as HTMLSlotElement

        const assignedElements = slot.assignedElements({
            flatten: true,
        }) as HTMLElement[]
        this.slots = assignedElements
        //@ts-ignore
        assignedElements.map((el, index) => {
            //@ts-ignore
            // el.trackId = ++index
            // el.setAttribute('track-id', 10)
        })
    }
    goToMin() {
        const marg = this.calcPlayheadFromTime('2021-02-01T01:30:00Z')

        this.margin = marg
    }
    render() {
        return (
            <Host>
                <div class="border">
                    {this.totalCol}
                    <button onClick={() => this.goToMin()}>go</button>
                    <div
                        class="rux-timeline"
                        ref={(el) => (this.slotContainer = el)}
                        onMouseMove={(ev) => this.handleMouse(ev)}
                        style={{
                            gridTemplateColumns: `[header] 200px repeat(${this.totalCol}, ${this.zoom}px)`,
                        }}
                    >
                        <div
                            class="rux-playhead"
                            style={{ left: `${this.margin}px` }}
                        ></div>
                        <slot onSlotchange={this._handleSlotChange}></slot>
                    </div>
                    <div>The current time is {this.time}</div>
                    {this.formattedCurrentTime}
                </div>
            </Host>
        )
    }
}
