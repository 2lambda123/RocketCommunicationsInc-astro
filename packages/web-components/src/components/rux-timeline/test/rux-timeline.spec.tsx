import { newSpecPage } from '@stencil/core/testing'
import { RuxTimeline } from '../rux-timeline'

describe('rux-timeline', () => {
    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxTimeline],
            html: `<rux-timeline start="2021-02-01T00:00:00Z" end="2021-02-05T12:00:00Z"></rux-timeline>`,
        })

        expect(page.root).toMatchSnapshot()
    })
})
