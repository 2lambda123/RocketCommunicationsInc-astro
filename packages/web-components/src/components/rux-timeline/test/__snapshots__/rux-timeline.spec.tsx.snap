// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-timeline renders 1`] = `
<rux-timeline end="2021-02-05T12:00:00Z" start="2021-02-01T00:00:00Z" zoom="1">
  <mock:shadow-root>
    <div class="rux-timeline" style="grid-template-columns: [header] 200px repeat(6480, 1px);">
      <slot></slot>
    </div>
  </mock:shadow-root>
</rux-timeline>
`;
